### ISSUE TITLE
YOLOv8 object detection not working after exporting to ONNX

### ISSUE THREAD

**Issue by kasperikaivola** [ultralytics/ultralytics#16346]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello! I've trained my custom YOLOv8 model starting from yolov8n.pt for detecting 33 classes of objects, tested the model performance after training for 200 epochs using the YOLO object (code below) It worked well with my test image, and I saved the model to an ONNX file. Now, I've tried to load the model from this ONNX file (best.onnx) using multiple different ways / code snippets and the result is always the same: the model doesn't detect anything (from the same test images as earlier) unless I set the thresholds extremely low (0.00001 or so), and even then only the bounding boxes are correct, not the object classes. Running the original yolov8n.onnx model using these same code snippets and pictures works perfectly.

I used this code to run the model: https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-OpenCV-ONNX-Python/main.py

Also tried this way:
```
# Load the exported ONNX model
onnx_model = YOLO("yolov8n.onnx")

# Run inference
results = onnx_model("./path/to/my/image")
```

Same result, what am I doing wrong?

Here is my training code:

```
def train_model(epochs=3):
    # Load a model
    model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

    # Use the model
    try:
        model.train(data=os.environ['YAMLFILE'], epochs=epochs, device=0)  # train the model
    except Exception as e:
        if "torch.cuda.is_available(): False" in repr(e):
            val = input("CUDA is not available in Torch, do you want to continue? (y/n): ")
            if val != 'y':
                return -1
        model.train(data=os.environ['YAMLFILE'], epochs=epochs)  # train the model using CPU
    metrics = model.val()  # evaluate model performance on the validation set
    path = model.export(format="onnx")  # export the model to ONNX format
    try:
        results = model("./dataset/test_images/7158485.jpg")  # predict on an image, I got accurate results here
        for r in results:
            print(r.boxes.data)
            print(r.names)
            idx = r.boxes.cls[0]
            print(idx)
            class_names = r.names[int(idx)]
            print(class_names)
    except Exception as e:
        print(e)
        print("Exception when trying to predict on an image, but model was still saved!")

if __name__ == "__main__":
    train_model()
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kasperikaivola, thank you for reaching out and using Ultralytics 🚀!

We recommend checking our comprehensive [Docs](https://docs.ultralytics.com) for insights, where many common questions might already be addressed. You can also explore specific usage examples for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/).

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and resolve the issue.

For questions regarding custom training ❓, kindly provide detailed information, including dataset examples and logs. Ensure you follow our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our community for more interaction. Engage in real-time discussions on [Discord](https://ultralytics.com/discord) 🎧, explore our [Discourse](https://community.ultralytics.com) for in-depth discussions, or dive into our [Subreddit](https://reddit.com/r/ultralytics) to share insights.

## Upgrade

Ensure you are using the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLOv8 can be utilized in these verified environments:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, ensuring correct operation of YOLOv8 across platforms.

This is an automated response; an Ultralytics engineer will be with you soon to assist further. 😊

**Comment by Haseeeb21** [comment]
Load the best.pt trained model. Export it in onnx format and try with half parameter set to `True`

model = YOLO("path/to/best.pt")

model.export(format="onnx", half=True)

then predict using

model.predict("path/to/img.jpg", save=True, conf=0.5)

or on real-time camera feed,

results = model.predict(source="0", show=True, conf=0.5)

Try it and do share the results.

**Comment by Skillnoob** [comment]
Make sure to have `onnxruntime_gpu` installed when using the half parameter as its for gpu use only. Ofc you also need a nvidia gpu.

**Comment by Haseeeb21** [comment]
> Make sure to have `onnxruntime_gpu` installed when using the half parameter as its for gpu use only. Ofc you also need a nvidia gpu.

Where is this written? can you share the source

**Comment by Skillnoob** [comment]
@Haseeeb21 ultralytics prints a warning when exporting with half and onnxruntime cpu

**Comment by kasperikaivola** [comment]
Thanks for the help @Skillnoob ! Saving the model to .pt and loading it again and converting to ONNX indeed fixed the issue. Here's the full code for anyone looking for a solution. I removed the model.export(format="onnx") from my training function and did this separately:
```
model = YOLO("best.pt")

model.export(format="onnx", simplify=True)
```
